### ISSUE TITLE
False-positive for immediate variable override

### ISSUE THREAD

**Issue by zerkms** [kalessil/phpinspectionsea#1342]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Extended) v3.0.13 |
| Language level |  PHP 7.2                     |

Current behaviour
--
```php
function a(): void
{
    try {
        $v = 0;
        $v = t();
    } finally {
        var_dump($v);
    }
}

function t(): int
{
    throw new \Exception('catch me if you can');
}

a();
```
It marks the second `$v = t();` assignment as "is immediately overridden".

Expected behaviour
--
Ideally, it should not

Environment details
---
N/A

**Comment by funivan** [comment]
Can you explain why not? Why we need this line in the code `$v=0`?

**Comment by zerkms** [comment]
@funivan run that code with and without `$v = 0;`

**Comment by funivan** [comment]
Oh, I got it. 
But you can rewrite tour code to this:
```php
 try {
        $v = t();
    } finally {
        var_dump(0);
    }
```
Behavior will be the same. Or i`m wrong?

**Comment by zerkms** [comment]
@funivan 

the whole example can be rewritten to 

```php
var_dump(0);
throw new \Exception('catch me if you can');
```

My report here is not about "please help me rewrite that piece of code", it's about this very detection that produces a false positive.

**Comment by funivan** [comment]
Analyzing exception inside functions is not so easy. If `t` is a simple function without any nested function calls then this is probably possible. But if `t` contains another function call then we need to analyze each nested function to check if it can throw an exception. Maybe kalessil can do this)

IMHO: this is a bug, but do not know if it is fixable ツ.

**Comment by zerkms** [comment]
Yep, I'd say without checked exceptions it's impossible to determine reliably.

It looks in general if there is a function call on the next line - it may throw. Not sure about operations.

Can operators throw in php? What other types of expressions other than operators and function calls can throw?

Possibly - everything if one converts warnings and notices to exceptions.

**Comment by funivan** [comment]
We can get type error exception in php 7.2 =)

**Comment by zerkms** [comment]
Not by default for everything https://3v4l.org/rkp7g

**Comment by kalessil** [comment]
Looks like a false-positive in this case.

**Comment by zerkms** [comment]
Yep, and after thinking on it for some time now I agree with @funivan that in general one simply should just avoid writing code like that (it may be not always trivial but still)

**Comment by kalessil** [comment]
Yep try should contain only statements which you want to guard, so the initial zero assignment ideally should be outside of the try-statement.

**Comment by zerkms** [comment]
The problem is that irl the initialisation is also dynamic and may throw :-D

**Comment by kalessil** [comment]
I suspected it'll be not so easy ;)

**Comment by kalessil** [comment]
Fixed!

**Comment by zerkms** [comment]
@kalessil 

![who is awesome](https://i.kym-cdn.com/photos/images/original/000/066/552/whos-awesome.jpg)
